src/solve/SimulationParameters.cpp

```diff
@@ -90,44 +198,74 @@ SimulationParameters load_simulation_params(const nlohmann::json& config) {
 
   // Create vessels
-  // DEBUG_MSG("Load vessels");
   std::map<int, std::string> vessel_id_map;
-  const auto& vessels = config["vessels"];
+  if (config.contains("vessels")) {
+    create_vessels(model, connections, config["vessels"], vessel_id_map);
+  }
+
+  // Create map for boundary conditions to boundary condition type
+  std::map<std::string, std::string> bc_type_map;
+  for (const auto& bc_config : config["boundary_conditions"]) {
+    std::string bc_name = bc_config["bc_name"];
+    std::string bc_type = bc_config["bc_type"];
+    bc_type_map.insert({bc_name, bc_type});
+  }
+
+  // Create external coupling blocks
+  if (config.contains("external_solver_coupling_blocks")) {
+    create_coupling(model, connections,
+                    config["external_solver_coupling_blocks"], vessel_id_map,
+                    bc_type_map);
+  }
+
+  // Create boundary conditions
+  std::vector<std::string> closed_loop_bcs;
+  create_bounary_conditions(model, config["boundary_conditions"], bc_type_map,
+                            closed_loop_bcs);
+
+  // Create junctions
+  if (config.contains("junctions")) {
+    create_junctions(model, connections, config["junctions"], vessel_id_map);
+  }
 
+  // Create closed-loop blocks
+  if (config.contains("closed_loop_blocks")) {
+    create_closed_loop(model, connections, config["closed_loop_blocks"],
+                       closed_loop_bcs);
+  }
+
+  // Create Connections
+  for (auto& connection : connections) {
+    auto ele1 = model.get_block(std::get<0>(connection));
+    auto ele2 = model.get_block(std::get<1>(connection));
+    model.add_node({ele1}, {ele2}, ele1->get_name() + ":" + ele2->get_name());
+  }
+
+  // Finalize model
+  model.finalize();
+}
+
+void create_vessels(
+    Model& model,
+    std::vector<std::tuple<std::string, std::string>>& connections,
+    const nlohmann::json& vessels, std::map<int, std::string>& vessel_id_map) {
+  // Loop all vessels
   for (size_t i = 0; i < vessels.size(); i++) {
     const auto& vessel_config = vessels[i];
     const auto& vessel_values = vessel_config["zero_d_element_values"];
     const std::string vessel_name = vessel_config["vessel_name"];
     vessel_id_map.insert({vessel_config["vessel_id"], vessel_name});
 
-    if (vessel_config["zero_d_element_type"] == "BloodVessel") {
-      model.add_block(BlockType::blood_vessel,
-                      {model.add_parameter(vessel_values["R_poiseuille"]),
-                       model.add_parameter(vessel_values.value("C", 0.0)),
-                       model.add_parameter(vessel_values.value("L", 0.0)),
-                       model.add_parameter(
-                           vessel_values.value("stenosis_coefficient", 0.0))},
-                      vessel_name);
-      // DEBUG_MSG("Created vessel " << vessel_name);
-    } else {
-      throw std::invalid_argument("Unknown vessel type");
-    }
+    generate_block(model, vessel_values, vessel_config["zero_d_element_type"],
```

Comment:
I'm confused that this call doesn't match the arguments in `generate_block` and I don't see an overloaded function that matches this call. Specifically, the second and third arguments here are the same (`vessel_values = vessel_config["zero_d_element_values"]`). But the second argument should be a json config and the third should be a string. Am I missing something?

Reply:
Argument two `vessel_values = vessel_config["zero_d_element_values"]` contains the `nlohmann::json` "sub-json" configuration of this block (e.g., `{C: 0.001, L: 0.01}`).
Argument three `vessel_config["zero_d_element_type"]` contains only the `std::string` type of the vessel (e.g., `BloodVessel`).

The confusion is probably that both arguments are extracted from the `json` configuration of this block. However, it does require two separate arguments since they look different in each of the `create_[...]` functions:
- `create_vessels`: `zero_d_element_values` and `zero_d_element_type`
- `create_boundary_conditions`: `bc_values` and `bc_type`
- `create_external_coupling`: `values` and `name`
- `create_junctions`: `junction_values` and `junction_type`
- ...

Maybe we could unify that in the future by changing our input files?

Reply:
Ah ok, sorry I accidentally read the third argument as `zero_d_element_values` instead of `zero_d_element_type`. Makes sense now!